### PR TITLE
[Playground] [Backend] Sharing any code

### PR DESCRIPTION
**Pull request by vchunikhin**
**Please** add a meaningful description for your change here
@konstantinurysov 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/start_datastore_emulator.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (VladMatyunin) [inline playground/backend/start_datastore_emulator.sh:1]: Please move the files for CI/CD to a separate PR

File: playground/backend/internal/db/schema/migration/initial_structure.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	if err = args.Db.PutSDKs(args.Ctx, sdkEntities); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type SdkConfig struct {
+	Sdks struct {
+		Go     *SdkProperties `yaml:"SDK_GO"`
```
Comment (VladMatyunin) [inline playground/backend/internal/db/schema/migration/initial_structure.go:92]: Please explore whether we can move that to a separate global struct

File: playground/backend/internal/db/local/local_db.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	l.items[id] = snip
+	return nil
+}
+
+// GetSnippet returns the snippet entity
+func (l *LocalDB) GetSnippet(_ context.Context, id string) (*entity.SnippetEntity, error) {
+	l.RLock()
+	value, found := l.items[id]
+	if !found {
+		l.RUnlock()
```
Comment (konstantinurysov) [inline playground/backend/internal/db/local/local_db.go:49]: what was the reason for not using defer here?

File: playground/backend/internal/db/local/local_db_test.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+var ctx context.Context
+var localDb *LocalDB
+
+func init() {
+	ctx = context.Background()
+	var err error
+	localDb, err = New()
+	if err != nil {
+		panic(err)
```
Comment (konstantinurysov) [inline playground/backend/internal/db/local/local_db_test.go:35]: please use panic only if it is inevitable
https://github.com/uber-go/guide/blob/master/style.md#dont-panic

File: playground/backend/internal/utils/file_utils.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	case pb.Sdk_SDK_JAVA:
+		return getCorrectNameOrDefault(ext, javaExt, defaultJavaFileName, name)
+	case pb.Sdk_SDK_GO:
+		return getCorrectNameOrDefault(ext, goExt, defaultGoFileName, name)
+	case pb.Sdk_SDK_PYTHON:
+		return getCorrectNameOrDefault(ext, pythonExt, defaultPythonFileName, name)
+	case pb.Sdk_SDK_SCIO:
+		return getCorrectNameOrDefault(ext, scioExt, defaultScioFileName, name)
+	}
+	return name
```
Comment (konstantinurysov) [inline playground/backend/internal/utils/file_utils.go:71]: nit: can be moved to default case

File: playground/backend/internal/utils/file_utils.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+// getCorrectNameOrDefault returns the correct file name or default name.
+func getCorrectNameOrDefault(actualExt, correctExt, defaultFileName, name string) string {
+	if actualExt == "" {
+		logger.Error("The name of the file does not have extension. Will be used default value")
+		return defaultFileName
+	}
+	if actualExt != correctExt {
+		logger.Error("The name of the file has wrong extension. Will be used correct extension according to sdk")
+		return name[0:len(name)-len(actualExt)] + correctExt
+	} else {
```
Comment (konstantinurysov) [inline playground/backend/internal/utils/file_utils.go:83]: nit: else can be removed

File: playground/backend/internal/utils/file_utils.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	case pb.Sdk_SDK_JAVA:
+		return strings.Contains(content, javaMainMethod)
+	case pb.Sdk_SDK_GO:
+		return strings.Contains(content, goMainMethod)
+	case pb.Sdk_SDK_PYTHON:
+		return strings.Contains(content, pythonMainMethod)
+	case pb.Sdk_SDK_SCIO:
+		return strings.Contains(content, scioMainMethod)
+	}
+	return false
```
Comment (konstantinurysov) [inline playground/backend/internal/utils/file_utils.go:100]: nit: can be moved to the default case

**Review by konstantinurysov — CHANGES_REQUESTED** [review]

File: playground/backend/cmd/server/controller_test.go [hunk 8]
```diff
@@ -80,6 +86,12 @@ func setup() *grpc.Server {
 	// setup cache
 	cacheService = local.New(context.Background())
 
+	// setup entity storage
+	snippetDb, err = localdb.New()
+	if err != nil {
+		panic(err)
```
Comment (konstantinurysov) [inline playground/backend/cmd/server/controller_test.go:92]: please use panic only if it is inevitable, just double check that we cannot replace it with other options
https://github.com/uber-go/guide/blob/master/style.md#dont-panic

**Pull request closed without merging**